### ISSUE TITLE
vmauth: doesnt skip unavailable vmselect/vminsert

### ISSUE THREAD

**Issue by vandud** [VictoriaMetrics/VictoriaMetrics#3061]
**Is your feature request related to a problem? Please describe.**
I have 3 nodes vmselect and vminsert. When some vmselect/vminsert is unavailable, vmauth sends queries to them anyway
In example below node-2 is stopped:
```
$ curl -s -I http://mon-vm-01:8481/select/1/prometheus/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-vm-01

$ curl -I http://mon-vm-02:8481/select/1/prometheus/api/v1/query?query=node_load1
curl: (7) Failed to connect to mon-vm-02 port 8481: Connection refused

$ curl -s -I http://mon-vm-03:8481/select/1/prometheus/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-vm-03
```
But vmauth anyway trying query it:
```
curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-proxy-01x
X-Server-Hostname: mon-vm-01

$ curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 502 Bad Gateway
X-Server-Hostname: mon-proxy-01x

$ curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-proxy-01x
X-Server-Hostname: mon-vm-03
```

**Describe the solution you'd like**
vmauth could skip unavailable url from url_prefix config section and move on to the next one url (if url_prefix contains more then one urls)

**Describe alternatives you've considered**
Now I using nginx after vmauth for solving this problem:
```
root@mon-proxy-01x:/etc/vmauth# cat main.yaml
---
users:
  - username: "tenant-1"
    password: "tenant-1"
    url_map:
      - src_paths:
          - "/api/v1/query"
          - "/api/v1/query_range"
          - "/api/v1/label/[^/]+/values"
        url_prefix:
          - "http://localhost:8428/select/1/prometheus"
      - src_paths:
          - "/api/v1/write"
        url_prefix:
          - "http://localhost:8428/insert/1/prometheus"
          - 
root@mon-proxy-01x:/etc/vmauth# cat /etc/nginx/sites-enabled/default
upstream vminsert {
        server mon-vm-01:8480;
        server mon-vm-02:8480;
        server mon-vm-03:8480;
}
upstream vmselect {
        server mon-vm-01:8481;
        server mon-vm-02:8481;
        server mon-vm-03:8481;
}
server {
  listen 8428;
  location /insert {
    proxy_pass http://vminsert;
  }
  location /select {
    proxy_pass http://vmselect;
  }
}
```
With nginx its working ok:
```
$ curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-proxy-01x
X-Server-Hostname: mon-vm-01

$ curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-proxy-01x
X-Server-Hostname: mon-vm-03

$ curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-proxy-01x
X-Server-Hostname: mon-vm-01

$ curl -s -I -u tenant-1:tenant-1 http://mon-proxy-01x:8427/api/v1/query?query=node_load1 | egrep --color=never 'HTTP|Hostname'
HTTP/1.1 200 OK
X-Server-Hostname: mon-proxy-01x
X-Server-Hostname: mon-vm-03
```
As You can see, nginx skips the unavailable mon-vm-02

**Comment by oubayun** [comment]
yes, this issue has been around.

**Comment by hasanawad94** [comment]
Still no solution for this ?

**Comment by nulldoot2k** [comment]
If you using **tenant-1:tenant-1** for multi vmagent then how can trace logs each service ?
For example vmauth:
```bash
---
users:
  - username: "tenant-1"
    password: "tenant-1"
    url_map:
      - src_paths:
          - "/api/v1/query"
          - "/api/v1/query_range"
          - "/api/v1/label/[^/]+/values"
        url_prefix:
          - "http://localhost:8481/select/1000/prometheus"
 ```

i want trace logs service users tenant-1

**Comment by recursive2018** [comment]
Any updates with regards to this issue? I can confirm it is still occurring on v1.105.0. 

Also unsure why it is tagged as an enhancement as the vmauth docs list loadbalancing as one of the features and the code appears to be doing checks to try and determine if the backend isBroken. Those checks don't seem to be working as requests still try to send to the backend vmselect/vminsert nodes in the case that the service is stopped and/or the backend is offline altogether.

Thanks.

**Comment by hagen1778** [comment]
@recursive2018 @vandud vmauth will try all the configured [HTTP backends](https://docs.victoriametrics.com/vmauth/#generic-http-load-balancer) to serve the user's query. If backend responds with an error, it is marked as `broken` for `-failTimeout` time interval. Once this interval passes, vmauth will try proxying new requests to this backend again, in case if it recovered. 

So it is expected for vmauth to send requests to unavailable backends from time to time. But for end user this behavior is transparent, user won't see retries or errors. If you'd like to extend the interval of backend recovery - bump `-failTimeout` value.

**Comment by recursive2018** [comment]
Hello @hagen1778, thank you for the guidance on the -failTimeout setting, unfortunately this setting does not seem to help the issue I am encountering. 

I have set failTimeout to 60s for testing. Despite this being set I continually see the retries and failures on every attempt to the backend that is down. 

In the case of the vminsert this does not appear to cause issues with the data insertion (no gaps in data graphs) as I assume my vmagent is automatically retrying and successfully inserting into the other node. 

In the case of the read path via Grafana however this does cause errors visible to the user. When refreshing a dashboard after taking down one of the nodes I see various panels report no data with read object errors. These no data errors occur on each refresh on various different panels and I can confirm in the vmauth journal that it is attempting to hit the down backend on every request despite the failTimeout of 60s. This to me indicates it is either failing to mark the down nodes as broken, or at the very least it is not respecting the failTimeout setting at all.

Excerpt from the vmauth journal showing it still attempting to send requests to the backend that is down (service is stopped on that node). Note this is just a small portion of the logs with errors from Grafana to the multitenant endpoint but I also see the errors on specific tenant selects from vmalert that I am running using prometheus-benchmark.
```
Nov 07 09:41:58 vmauth001 vmauth-prod[30847]: 2024-11-07T15:41:58.482Z        warn        VictoriaMetrics/app/vmauth/main.go:291        remoteAddr: "10.10.10.10:47738"; requestURI: http://vmselect001:8481/select/multitenant/prometheus/api/v1/query_range; retrying the request to http://vmselect001:8481/select/multitenant/prometheus/api/v1/query_range because of response error: dial tcp4 10.10.10.11:8481: connect: connection refused; try -enableTCP6 command-line flag for dialing ipv6 addresses

Nov 07 09:42:00 vmauth001 vmauth-prod[30847]: 2024-11-07T15:42:00.587Z        warn        VictoriaMetrics/app/vmauth/main.go:291        remoteAddr: "10.10.10.10:47750"; requestURI: http://vmselect001:8481/select/multitenant/prometheus/api/v1/query_exemplars; retrying the request to http://vmselect001:8481/select/multitenant/prometheus/api/v1/query_exemplars because of response error: dial tcp4 10.10.10.11:8481: connect: connection refused; try -enableTCP6 command-line flag for dialing ipv6 addresses

Nov 07 09:42:00 vmauth001 vmauth-prod[30847]: 2024-11-07T15:42:00.657Z        warn        VictoriaMetrics/app/vmauth/main.go:291        remoteAddr: "10.10.10.10:47746"; requestURI: http://vmselect001:8481/select/multitenant/prometheus/api/v1/query; retrying the request to http://vmselect001:8481/select/multitenant/prometheus/api/v1/query because of response error: dial tcp4 10.10.10.11:8481: connect: connection refused; try -enableTCP6 command-line flag for dialing ipv6 addresses
```

Screenshot of the grafana error.
![image](https://github.com/user-attachments/assets/cc7a798c-6ce9-4e80-9bed-2e4d2d6a02d6)

Thank you.

**Comment by hagen1778** [comment]
Thanks for detailed comment @recursive2018 !
I think I found the issue - see https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7489.

**Comment by recursive2018** [comment]
@hagen1778 I was able to compile the updated vmauth from master and have confirmed I no longer see the issue. It attempts to reach the node once, fails to do so, logs it in the vmauth log and does not attempt to proxy requests to the node until the failTimeout is reached. None of the errors show in Grafana either.

Thanks for the quick turnaround!

**Comment by f41gh7** [comment]
The bugfix was included into [v1.106.1](https://docs.victoriametrics.com/changelog/#v11061), [v1.102.7](https://docs.victoriametrics.com/changelog/#v11027) and [v1.97.12](https://docs.victoriametrics.com/changelog/#v19712).
